auth/auth.go

```diff
@@ -0,0 +1,53 @@
+package auth
+
+import (
+	"strings"
+
+	"gopkg.in/src-d/go-vitess.v1/mysql"
+)
+
+// Permission holds permissions required by a query or grated to a user.
+type Permission int
+
+const (
+	// ReadPerm means that it reads.
+	ReadPerm Permission = 1 << iota
+	// WritePerm means that it writes.
+	WritePerm
+)
+
+var (
+	// AllPermissions hold all defined permissions.
+	AllPermissions = ReadPerm | WritePerm
+	// DefaultPermissions are the permissions granted to a user if not defined.
+	DefaultPermissions = ReadPerm
+
+	// PermissionNames is used to translate from human and machine
+	// representations.
+	PermissionNames = map[string]Permission{
+		"read":  ReadPerm,
+		"write": WritePerm,
+	}
+
+	// ErrNoPermission is returned when the user lacks needed permissions.
+	ErrNoPermission = "user does not have permission: %s"
+)
+
+// String returns all the permissions set to on.
+func (p Permission) String() string {
+	var str []string
+	for k, v := range PermissionNames {
+		if p&v != 0 {
+			str = append(str, k)
+		}
+	}
+
+	return strings.Join(str, ", ")
+}
+
+// Auth interface provides mysql authentication methods and permission checking
+// for users.
+type Auth interface {
+	Mysql() mysql.AuthServer
+	Allowed(user string, permission Permission) error
```

Comment:
Maybe return `(bool, error)`?

This would allow to differentiate from not having permissions and failing to check permissions. The former would probably produce a warning audit log (when we have audit logs), the second would probably also produce a regular error log.

Reply:
@kuba-- It's a common pattern in Go, also in our own codebase. `(true, err)` is usually not relevant since the value would not be even checked if `err != nil`.

But alternatively, you can keep just `err` as return value and use a special error kind `ErrNotAuthorized` to differentiate from other errors.

Reply:
@smola - totally understand, just as I mentioned, personally I don't like this _pattern_, because have a feeling that it's a _boolean logic with extra dimension_.

Reply:
I'm moving to use `go-errors` and returning `ErrNotAuthorized` so is easier to tell apart.